### PR TITLE
TST: ro_replace changing dtype of series

### LINKED ISSUE #25797: Different behavior of to_replace method between pandas version 0.23.4 and 0.24.2 by changing dtype of series
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

# fails in version 0.24.2 but works in 0.23.4
df = pd.DataFrame.from_dict({'Test': ['0.5', True, '0.6']})
print(df['Test'].dtype)
assert df['Test'].dtype == 'object'
df['Test'] = df['Test'].replace([True], [np.nan])
print(df['Test'].dtype)
assert df['Test'].dtype == 'object'

# fails in version 0.24.2 but works in 0.23.4
df = pd.DataFrame.from_dict({'Test': ['0.5', None, '0.6']})
print(df['Test'].dtype)
assert df['Test'].dtype == 'object'
df['Test'] = df['Test'].replace([None], [np.nan])
print(df['Test'].dtype)
assert df['Test'].dtype == 'object'

# works in both mentioned versions
df = pd.DataFrame.from_dict({'Test': ['0.5', None, '0.6']})
print(df['Test'].dtype)
assert df['Test'].dtype == 'object'
df['Test'] = df['Test'].fillna(np.nan)
print(df['Test'].dtype)
assert df['Test'].dtype == 'object'

```
#### Problem description

I have encountered this change of behavior from pandas version 0.23.4 to 0.24.2, but could not find anything in the release notes or the latest documentation, which would inform the user about this change. I am not sure whether the behavioral change of the method is an improvement or not. However, I think it should be mentioned somehow. In addition, since it seems that it is not mentioned, it may hint to a deeper problem of the to_replace method. 

For documentation-related issues, you can check the latest versions of the docs on `master` here:

https://pandas-docs.github.io/pandas-docs-travis/

#### Expected Output

# no assertion errors, dtype of column stays object for all three cases

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.2
pytest: 4.3.1
pip: 19.0.3
setuptools: 40.8.0
Cython: None
numpy: 1.15.4
scipy: 1.1.0
pyarrow: 0.12.1
xarray: None
IPython: 7.3.0
sphinx: 1.8.5
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.1
xlrd: 1.2.0
xlwt: None
xlsxwriter: 1.1.5
lxml.etree: 4.3.2
bs4: None
html5lib: None
sqlalchemy: 1.3.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by pooja-subramaniam**
- closes #25797
- Tests added and passed
@phofl and @noatamir

### CODE REVIEW

**Comment by noatamir** [comment]
👋  @pooja-subramaniam 
The pre-commit errors will be fixed if you:
1. Replace your single quotes with double quotes
2. Add a new line at the end of the file (after your test)

If you want, you can have a look at the details link to see what the errors look like, so you can identify them next time 😉.

**Comment by pooja-subramaniam** [comment]
thanks for the points @noatamir!

**Comment by pooja-subramaniam** [comment]
@noatamir I don't understand the reason for the canceled check here. Do you know why that happened?

File: pandas/tests/series/methods/test_replace.py [hunk 1]
```diff
@@ -675,3 +675,26 @@ def test_replace_value_none_dtype_numeric(self, val):
         result = ser.replace(val, None)
         expected = pd.Series([1, None], dtype=object)
         tm.assert_series_equal(result, expected)
+
+    def test_replace_change_dtype_series(self):
+        # GH#25797
+        df = pd.DataFrame.from_dict({"Test": ["0.5", True, "0.6"]})
+        print(df["Test"].dtype)
```
Comment (phofl) [inline pandas/tests/series/methods/test_replace.py:682]: Some pointers: Please remove the print statements.

Generally, you can remove your first assert statements per block, but we want to compare the whole df with the second assert statement, e.g. define your expected object and then call assert_frame_equal

```
expected = ...
tm.assert_frame_equal(df, expected)
```
 
if expected is the same as your initial df, you can define expected through

```
expected = df.copy()
```

in the beginning

**Review by phofl — COMMENTED** [review]
You can ignore cancelled pipelines if there is only one of them

This happens once in a while

File: pandas/tests/series/methods/test_replace.py (hunk 1 above)
Comment (pooja-subramaniam) [reply pandas/tests/series/methods/test_replace.py]: Thanks! Made the changes. 

Correct me if I am wrong, but here the assert needs to check if the series dtype remains 'object' and not if the dataframes are equal. I've made changes accordingly.

Comment (phofl) [reply pandas/tests/series/methods/test_replace.py]: tm.assert_frame_equal raises if your dtypes are different. So this is covered as well. In general, we want to check that the DataFrame is as expected (values, columns, index, dtype, ...). That is why we are using assert_frame_equal

The functions could modify the values unexpectedly and the test would not fail, if you only check the dtypes

Comment (pooja-subramaniam) [reply pandas/tests/series/methods/test_replace.py]: Ah okay! Thanks for explaining :)

File: pandas/tests/series/methods/test_replace.py [hunk 2]
```diff
@@ -675,3 +675,20 @@ def test_replace_value_none_dtype_numeric(self, val):
         result = ser.replace(val, None)
         expected = pd.Series([1, None], dtype=object)
         tm.assert_series_equal(result, expected)
+
+    def test_replace_change_dtype_series(self):
+        # GH#25797
+        df = pd.DataFrame.from_dict({"Test": ["0.5", True, "0.6"]})
+        df["Test"] = df["Test"].replace([True], [np.nan])
+        expected = pd.DataFrame.from_dict({"Test": ["0.5", np.nan, "0.6"]})
```
Comment (phofl) [inline pandas/tests/series/methods/test_replace.py:683]: One last comment: Could you define expected only once after line 681 and reuse it for all assert_frame_equal statements? It's not necessary to define it multiple times when it's the same object :)

Comment (pooja-subramaniam) [reply pandas/tests/series/methods/test_replace.py:683]: Right! How did I miss that one! Thanks again ;)

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @pooja-subramaniam

**Pull request merged**